gapic/schema/api.py

```diff
@@ -560,6 +561,119 @@ def mixin_http_options(self):
+                    methods[f"{service_key}.{method_key}"] = method_value
+            return methods
+
+        methods = _get_all_methods(self)
+
+        def _get_auto_populated_fields_candidates(
+            message: MethodDescriptorProto,
+        ) -> Sequence[str]:
+            """Return fields which meet the criteria to be automatically populated for a given message.
+            For a field to be automatically populated, all the below configurations must
+            be true according to https://google.aip.dev/client-libraries/4235:
+
+            - The field must be of type string
+            - The field must be at the top-level of the request message
+            - The field must not be annotated with google.api.field_behavior = REQUIRED.
+            - The field must be annotated with google.api.field_info.format = UUID4.
+
+            Args:
+                message[MethodDescriptorProto]: Describes a method of a service.
+            Return:
+                Sequence[str]: A list of candidates to be automatically populated
+                    according to https://google.aip.dev/client-libraries/4235.
+            """
+            top_level_request_message = self.messages[
+                message.input_type.lstrip(".")
+            ]
+            return [
+                field.name
+                for field in top_level_request_message.fields.values()
+                if field.type == wrappers.PrimitiveType.build(str)
+                and not field.required
+                and field.uuid4
+            ]
+
+        def _get_auto_populated_fields(
+            method_settings: client_pb2.MethodSettings,
+        ) -> Sequence[str]:
+            """
+            Return a list of auto-populated fields based on the service YAML.
+
+            Args:
+                method_settings (client_pb2.MethodSettings): The settings
+                to be used when generating API methods.
+            Return:
+                Sequence[str]: A sequence of all fields to be automatically
+                    populated, taking into account the requirements
+                    from AIP-4235.
+                    https://google.aip.dev/client-libraries/4235
+
+            Raises:
+                ValueError: if `auto_populated_fields` are incorrectly set in the service config
+            """
+            try:
+                method_in_scope = methods[method_settings.selector]
+            except KeyError:
+                raise ValueError(
+                    f"Selector {method_settings.selector} is not a valid method in this API."
+                )
+
+            if (
+                method_in_scope.client_streaming or method_in_scope.server_streaming
+            ) and method_settings.auto_populated_fields:
+                raise ValueError(
+                    f"Selector {method_settings.selector} is a streaming rpc. `auto_populated_fields` are only supported in unary rpcs."
+                )
```

Comment:
My suggestion would be to split up this function and have a separate function for each check.

```python3
def is_streaming_rpc(method_in_scope)
    if ( 
    method_in_scope.client_streaming or
    method_in_scope.server_streaming
    ) and
    method_settings.auto_populated_fields:
      raise ValueError(...)
    return False
```

Overall I do understand how the code flow works, but I find it hard to follow through the tests due to the nested functions. Not necessary, but my suggestion would be to move these functions out of `get_all_methods` and test each piece individually which in my opinion will be much easier to follow through!
Similarly, for the code below, we could have:
```python3
def is_top_level_method():
   # add logic to check top level field
```

Reply:
I refactored the code in https://github.com/googleapis/gapic-generator-python/pull/1975/commits/41c582c6a1f87481b94b237bc76a65eea0273b0c. Please take another look